hack/tools/release/notes/list.go

```diff
@@ -0,0 +1,143 @@
+//go:build tools
+// +build tools
+
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// githubFromToPRLister lists PRs from GitHub contained between two refs.
+type githubFromToPRLister struct {
+	client         *githubClient
+	fromRef, toRef ref
+	// branch is optional. It helps optimize the PR query by restricting
+	// the results to PRs merged in the selected branch and in main
+	branch string
+}
+
+func newGithubFromToPRLister(repo string, fromRef, toRef ref, branch string) *githubFromToPRLister {
+	return &githubFromToPRLister{
+		client:  &githubClient{repo: repo},
+		fromRef: fromRef,
+		toRef:   toRef,
+		branch:  branch,
+	}
+}
+
+// listPRs returns the PRs merged between `fromRef` and `toRef` (included).
+func (l *githubFromToPRLister) listPRs() ([]pr, error) {
+	log.Printf("Computing diff between %s and %s", l.fromRef, l.toRef)
+	diff, err := l.client.getDiffAllCommits(l.fromRef.value, l.toRef.value)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Reading ref %s for upper limit", l.toRef)
+	toRef, err := l.client.getRef(l.toRef.String())
+	if err != nil {
+		return nil, err
+	}
+
+	var toCommitSHA string
+	if toRef.Object.ObjectType == tagType {
+		log.Printf("Reading tag info %s for upper limit", toRef.Object.SHA)
+		toTag, err := l.client.getTag(toRef.Object.SHA)
+		if err != nil {
+			return nil, err
+		}
+		toCommitSHA = toTag.Object.SHA
+	} else {
+		toCommitSHA = toRef.Object.SHA
+	}
+
+	log.Printf("Reading commit %s for upper limit", toCommitSHA)
+	toCommit, err := l.client.getCommit(toCommitSHA)
+	if err != nil {
+		return nil, err
+	}
+
+	fromDate := diff.MergeBaseCommit.Commit.Committer.Date
+	// We add an extra minute to avoid errors by 1 (observed during testing)
+	// We cross check the list of PRs against the list of commits, so we will filter out
+	// any PRs entries not belonging to the computed diff
+	toDate := toCommit.Committer.Date.Add(1 * time.Minute)
+
+	log.Printf("Listing PRs from %s to %s", fromDate, toDate)
+	gPRs, err := l.client.listMergedPRs(fromDate, toDate, l.branch, "main")
```

Comment:
Does this work for PRs merged into release branches? (as we're hard-coding main as the base branch)

Reply:
it does work yeah, that's what `l.branch` is (it lists PRs both to main and to `l.branch`)

We include `main` here because when cutting a new minor version, there will PRs merged in both main and the release branch. The returned PRs from this list are filtered from the list of commits in between the refs `to` and `from`, so it won't print PRs from `main` that belong to a different release, even if merged during the same date range. 

This is just an optimization to avoid listing PRs over all branches, since we know we don't need the PRs from other release branches.

Reply:
Sounds good. Can you add a godoc comment explaingin this? Thank you :)

Reply:
Done :)